atoma-service/src/middleware.rs

```diff
@@ -443,119 +449,182 @@ pub async fn verify_stack_permissions(
+                endpoint: endpoint.clone(),
+            })?
+            .parse::<i64>()
+            .map_err(|e| AtomaServiceError::InvalidHeader {
+                message: format!("Stack small ID is not a valid integer, with error: {e}"),
+                endpoint: endpoint.clone(),
+            })?;
+        let (result_sender, result_receiver) = oneshot::channel();
+        state
+            .state_manager_sender
+            .send(
+                AtomaAtomaStateManagerEvent::GetAvailableStackWithComputeUnits {
+                    stack_small_id,
+                    sui_address: sui_address.to_string(),
+                    total_num_compute_units: max_total_compute_units,
+                    result_sender,
+                },
             )
-            .await?;
-            // NOTE: We do not need to check that the stack small id matches the one in the request,
-            // or that the number of compute units within the stack is enough for processing the request,
-            // as the Sui subscriber service should handle this verification.
-        }
-        StackAvailability::Locked => {
-            // NOTE: If we are within this branch logic, it means that there is a stack with the same
-            // stack_small_id, but it is locked, so the user needs to buy a new stack, and we provide
-            // a specific status code to flag this scenario to the client.
-            return Err(AtomaServiceError::LockedStackError {
-                message: format!(
-                    "Stack with stack_small_id={stack_small_id} is locked, please buy a new stack."
+            .map_err(|err| AtomaServiceError::InternalError {
+                message: format!("Failed to get available stacks: {}", err),
+                endpoint: endpoint.clone(),
+            })?;
+        let available_stack = result_receiver
+            .await
+            .map_err(|e| AtomaServiceError::AuthError {
+                auth_error: format!(
+                    "Failed to get available stack with enough compute units, with error: {e}"
                 ),
                 endpoint: endpoint.clone(),
-            });
-        }
-        StackAvailability::Unavailable => {
-            // NOTE: If we are within this branch logic, it means that there is a stack with the same
-            // stack_small_id, but it is unavailable, so the client either buys a new stack or awaits
-            // the stack to be available again.
-            return Err(AtomaServiceError::UnavailableStackError {
-                message: format!(
-                    "Stack with stack_small_id={stack_small_id} is unavailable, please buy a new stack or await it to be available again."
+            })?
+            .map_err(|err| AtomaServiceError::AuthError {
+                auth_error: format!(
+                    "Failed to get available stack with enough compute units, with error: {err}"
                 ),
                 endpoint: endpoint.clone(),
+            })?;
```

Comment:
All this logic could be encapsulated in a function,  and we should have one for fiat and stack payments.

Reply:
done